src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7743,11 +7762,16 @@ private ForStatementSyntax ParseForStatement()
                 }
                 else
                 {
-                    isDeclaration = !this.IsQueryExpression(mayBeVariableDeclaration: true, mayBeMemberDeclaration: false) &&
-                                    this.ScanType() != ScanTypeFlags.NotType &&
-                                    this.IsTrueIdentifier();
+                    decl = TryParseDeconstructionDeclaration(withEquals: true);
 
-                    this.Reset(ref resetPoint);
+                    if (decl == null)
+                    {
+                        isDeclaration = !this.IsQueryExpression(mayBeVariableDeclaration: true, mayBeMemberDeclaration: false) &&
+                                        this.ScanType() != ScanTypeFlags.NotType &&
+                                        this.IsTrueIdentifier();
+
+                        this.Reset(ref resetPoint);
+                    }
```

Comment:
If you mean the case where `decl == null`, I added one. That is simply ` for(Int32 x = foo; ; ) { }` which I verified enters line 7769.